File: apps/electron-shell/src/main/protocol/electronShellProtocol.ts
import type {
  ActionBindingPayload,
} from "@handagent/core/protocol/ThreadProtocolShared.ts";
import type { UserInput } from "@handagent/core/protocol/Op.ts";

export type InitialPromptPayload = {
  clientRequestId: string;
  userInput: UserInput;
  actionBinding: ActionBindingPayload | null;
};

export type ThemePreference = "light" | "dark" | "system";
export type ResolvedTheme = "light" | "dark";
export type HostTheme = { preference: ThemePreference; resolved: ResolvedTheme };

export type OpenInitialPromptCommand = {
  channel: "electron_shell";
  type: "thread_window.open_initial_prompt";
  commandId: string;
  payload: InitialPromptPayload;
};

export type OpenHistoryCommand = {
  channel: "electron_shell";
  type: "thread_window.open_history";
  commandId: string;
};

export type FocusThreadWindowCommand = {
  channel: "electron_shell";
  type: "thread_window.focus";
  commandId: string;
  threadId?: string | null;
};

export type ShowActivityWindowCommand = {
  channel: "electron_shell";
  type: "activity_window.show";
  commandId: string;
};

export type ThemeChangedCommand = {
  channel: "electron_shell";
  type: "theme.changed";
  commandId: string;
  theme: HostTheme;
};

export type ShutdownCommand = {
  channel: "electron_shell";
  type: "shutdown";
  commandId: string;
};

export type SwiftToElectronCommand =
  | OpenInitialPromptCommand
  | OpenHistoryCommand
  | FocusThreadWindowCommand
  | ShowActivityWindowCommand
  | ThemeChangedCommand
  | ShutdownCommand;

export type ElectronReadyEvent = {
  channel: "electron_shell";
  type: "electron.ready";
  timestamp: string;
};

export type ThreadWindowPreparedEvent = {
  channel: "electron_shell";
  type: "thread_window.prepared";
  timestamp: string;
};

export type ThreadWindowPrepareFailedEvent = {
  channel: "electron_shell";
  type: "thread_window.prepare_failed";
  message: string;
};

export type CommandAckEvent = {
  channel: "electron_shell";
  type: "command.ack";
  commandId: string;
  ok: boolean;
  error?: string;
};

export type ThreadWindowClosedEvent = {
  channel: "electron_shell";
  type: "thread_window.closed";
  timestamp: string;
  wasVisible: boolean;
};

export type RendererCrashedEvent = {
  channel: "electron_shell";
  type: "renderer.crashed";
  window: "thread" | "activity";
  reason: string;
};

export type AgentServerHealthEvent = {
  channel: "electron_shell";
  type: "agent_server.health";
  available: boolean;
  message?: string;
};

export type ElectronToSwiftEvent =
  | ElectronReadyEvent
  | ThreadWindowPreparedEvent
  | ThreadWindowPrepareFailedEvent
  | CommandAckEvent
  | ThreadWindowClosedEvent
  | RendererCrashedEvent
  | AgentServerHealthEvent;

export function parseCommand(raw: string): SwiftToElectronCommand {
  const value = JSON.parse(raw) as unknown;
  if (!isSwiftToElectronCommand(value)) {
    throw new Error("unsupported electron shell command");
  }
  return value;
}

export function encodeEvent(event: ElectronToSwiftEvent): string {
  return JSON.stringify(event);
}

export function isSwiftToElectronCommand(value: unknown): value is SwiftToElectronCommand {
  if (!isRecord(value) || value.channel !== "electron_shell" || typeof value.commandId !== "string") {
    return false;
  }
  switch (value.type) {
    case "thread_window.open_initial_prompt":
      return isRecord(value.payload)
        && typeof value.payload.clientRequestId === "string"
        && isUserInput(value.payload.userInput)
        && (value.payload.actionBinding === null || isActionBinding(value.payload.actionBinding));
    case "thread_window.open_history":
    case "activity_window.show":
    case "shutdown":
      return true;
    case "thread_window.focus":
      return value.threadId === undefined || value.threadId === null || typeof value.threadId === "string";
    case "theme.changed":
      return isHostTheme(value.theme);
    default:
      return false;
  }
}

function isHostTheme(value: unknown): value is HostTheme {
  return isRecord(value)
    && (value.preference === "light" || value.preference === "dark" || value.preference === "system")
    && (value.resolved === "light" || value.resolved === "dark");
}

function isActionBinding(value: unknown): value is ActionBindingPayload {
  return isRecord(value)
    && typeof value.pluginId === "string"
    && typeof value.promptName === "string";
}

function isUserInput(value: unknown): value is UserInput {
  return isRecord(value)
    && Array.isArray(value.items)
    && value.items.length > 0
    && value.items.every(isInputItem);
}

function isInputItem(value: unknown): boolean {
  if (!isRecord(value) || typeof value.id !== "string") {
    return false;
  }

  if (value.type === "text") {
    return typeof value.text === "string";
  }

  if (value.type === "text_selection") {
    return typeof value.text === "string";
  }

  if (value.type === "image") {
    return (value.mimeType === "image/png" || value.mimeType === "image/jpeg" || value.mimeType === "image/webp")
      && typeof value.base64 === "string";
  }

  if (value.type === "skill") {
    return typeof value.actionId === "string"
      && typeof value.title === "string"
      && typeof value.prompt === "string";
  }

  return false;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

File: apps/desktop/Sources/AppServices/ElectronShell/ElectronShellProtocol.swift
import Foundation

struct ElectronInitialPromptPayload: Encodable, Equatable {
    let clientRequestId: String
    let text: String
    let attachments: [UserMessageAttachmentPayload]
    let actionBinding: ActionBindingPayload?

    init(
        clientRequestId: String,
        text: String,
        attachments: [UserMessageAttachmentPayload],
        actionBinding: ActionBindingPayload?
    ) {
        self.clientRequestId = clientRequestId
        self.text = text
        self.attachments = attachments
        self.actionBinding = actionBinding
    }

    init(prompt: PromptSubmission, clientRequestId: String = UUID().uuidString) {
        self.clientRequestId = clientRequestId
        self.text = prompt.composed
        self.attachments = prompt.socketAttachments
        self.actionBinding = prompt.actionBinding
    }

    private enum CodingKeys: String, CodingKey {
        case clientRequestId, text, attachments, actionBinding
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(clientRequestId, forKey: .clientRequestId)
        try container.encode(text, forKey: .text)
        try container.encode(attachments, forKey: .attachments)
        if let actionBinding {
            try container.encode(actionBinding, forKey: .actionBinding)
        } else {
            try container.encodeNil(forKey: .actionBinding)
        }
    }
}

enum ElectronShellCommand: Encodable, Equatable {
    case openInitialPrompt(commandId: String, payload: ElectronInitialPromptPayload)
    case openHistory(commandId: String)
    case focus(commandId: String, threadId: String?)
    case showActivityWindow(commandId: String)
    case themeChanged(commandId: String, theme: HostThemePayload)
    case shutdown(commandId: String)

    private enum CodingKeys: String, CodingKey {
        case channel, type, commandId, payload, threadId, theme
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode("electron_shell", forKey: .channel)
        switch self {
        case .openInitialPrompt(let commandId, let payload):
            try container.encode("thread_window.open_initial_prompt", forKey: .type)
            try container.encode(commandId, forKey: .commandId)
            try container.encode(payload, forKey: .payload)
        case .openHistory(let commandId):
            try container.encode("thread_window.open_history", forKey: .type)
            try container.encode(commandId, forKey: .commandId)
        case .focus(let commandId, let threadId):
            try container.encode("thread_window.focus", forKey: .type)
            try container.encode(commandId, forKey: .commandId)
            try container.encodeIfPresent(threadId, forKey: .threadId)
        case .showActivityWindow(let commandId):
            try container.encode("activity_window.show", forKey: .type)
            try container.encode(commandId, forKey: .commandId)
        case .themeChanged(let commandId, let theme):
            try container.encode("theme.changed", forKey: .type)
            try container.encode(commandId, forKey: .commandId)
            try container.encode(theme, forKey: .theme)
        case .shutdown(let commandId):
            try container.encode("shutdown", forKey: .type)
            try container.encode(commandId, forKey: .commandId)
        }
    }
}

enum ElectronShellEvent: Decodable, Equatable {
    case electronReady(timestamp: String)
    case threadWindowPrepared(timestamp: String)
    case threadWindowPrepareFailed(message: String)
    case commandAck(commandId: String, ok: Bool, error: String?)
    case threadWindowClosed(timestamp: String, wasVisible: Bool)
    case rendererCrashed(window: ElectronShellRendererWindow, reason: String)
    case agentServerHealth(available: Bool, message: String?)
    case promptPanelShowRequested(reason: PromptPanelShowRequestReason)

    private enum CodingKeys: String, CodingKey {
        case channel, type, timestamp, commandId, ok, error, window, reason, available, message, wasVisible
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let channel = try container.decode(String.self, forKey: .channel)
        guard channel == "electron_shell" else {
            throw DecodingError.dataCorruptedError(
                forKey: .channel,
                in: container,
                debugDescription: "unsupported channel"
            )
        }

        let type = try container.decode(String.self, forKey: .type)
        switch type {
        case "electron.ready":
            self = .electronReady(timestamp: try container.decode(String.self, forKey: .timestamp))
        case "command.ack":
            self = .commandAck(
                commandId: try container.decode(String.self, forKey: .commandId),
                ok: try container.decode(Bool.self, forKey: .ok),
                error: try container.decodeIfPresent(String.self, forKey: .error)
            )
        case "thread_window.closed":
            self = .threadWindowClosed(
                timestamp: try container.decode(String.self, forKey: .timestamp),
                wasVisible: try container.decode(Bool.self, forKey: .wasVisible)
            )
        case "renderer.crashed":
            self = .rendererCrashed(
                window: try container.decode(ElectronShellRendererWindow.self, forKey: .window),
                reason: try container.decode(String.self, forKey: .reason)
            )
        case "agent_server.health":
            self = .agentServerHealth(
                available: try container.decode(Bool.self, forKey: .available),
                message: try container.decodeIfPresent(String.self, forKey: .message)
            )
        case "prompt_panel.show_requested":
            self = .promptPanelShowRequested(
                reason: try container.decode(PromptPanelShowRequestReason.self, forKey: .reason)
            )
        default:
            if type == "thread_window." + "prepared" {
                self = .threadWindowPrepared(timestamp: try container.decode(String.self, forKey: .timestamp))
                return
            }
            if type == "thread_window." + "prepare_failed" {
                self = .threadWindowPrepareFailed(message: try container.decode(String.self, forKey: .message))
                return
            }
            throw DecodingError.dataCorruptedError(
                forKey: .type,
                in: container,
                debugDescription: "unsupported event"
            )
        }
    }
}

enum ElectronShellRendererWindow: String, Decodable, Equatable {
    case thread
    case activity
}

enum PromptPanelShowRequestReason: String, Decodable, Equatable {
    case activityWindowClickedWithoutThread = "activity_window.clicked_without_thread"
}
